crud/src/main/java/com/redhat/lightblue/assoc/QueryPlanChooser.java

```diff
@@ -184,34 +181,44 @@ private void iterateReferences(CompositeMetadata root,List<Conjunct> unassignedC
      * nodes. These are fixed assignments, and don't change from one
      * query plan to the other
      */
-    private void assignQueriesToPlanNodesAndEdges(List<Conjunct> queries,List<Conjunct> unassigned) {
+    private void assignQueriesToPlanNodesAndEdges(List<Conjunct> queries,
+                                                  List<Conjunct> unassigned) {
         Error.push("assignQueriesToPlanNodesAndEdges");
         LOGGER.debug("Assigning queries to query plan nodes and edges");
         try {
             for(Conjunct c:queries) {
-                Set<QueryPlanNode> nodes=c.getReferredNodes();
+                Set<CompositeMetadata> entities=c.getEntities();
                 LOGGER.debug("Conjunct {}",c);
-                switch(nodes.size()) {
+                switch(entities.size()) {
                 case 1:
                     // All query clauses that depend on fields from a single
                     // entity can be assigned to a query plan node. If clauses
                     // depend on multiple entities, their assignments may change
                     // based on the query plan.
                     LOGGER.debug("Conjunct has one entity");
-                    nodes.iterator().next().getData().getConjuncts().add(c);
+                    // Rewrite the query for that node
+                    QueryPlanNode node=qplan.getNode(entities.iterator().next());
+                    RewriteQuery rw=new RewriteQuery(compositeMetadata,node.getMetadata());
+                    QueryExpression q=rw.rewriteQuery(c.getClause(),c.getFieldInfo()).query;
+                    AnalyzeQuery analyzer=new AnalyzeQuery(node.getMetadata(),c.getReference());
+                    analyzer.iterate(q);
+                    node.getData().getConjuncts().add(new Conjunct(q,
+                                                                   analyzer.getFieldInfo(),
+                                                                   c.getReference()));
                     break;
 
                 case 2:
                     // There are two or more entities referred to in the conjunct
                     // This clause can be associated with an edge
```

Comment:
Added more comments.

On Tue, Mar 29, 2016 at 4:59 AM, Dennis Crissman <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/assoc/QueryPlanChooser.java
> <https://github.com/lightblue-platform/lightblue-core/pull/608#discussion_r57705027>
> :
>
> >                  case 1:
> >                      // All query clauses that depend on fields from a single
> >                      // entity can be assigned to a query plan node. If clauses
> >                      // depend on multiple entities, their assignments may change
> >                      // based on the query plan.
> >                      LOGGER.debug("Conjunct has one entity");
> > -                    nodes.iterator().next().getData().getConjuncts().add(c);
> > +                    // Rewrite the query for that node
> > +                    QueryPlanNode node=qplan.getNode(entities.iterator().next());
> > +                    RewriteQuery rw=new RewriteQuery(compositeMetadata,node.getMetadata());
> > +                    QueryExpression q=rw.rewriteQuery(c.getClause(),c.getFieldInfo()).query;
> > +                    AnalyzeQuery analyzer=new AnalyzeQuery(node.getMetadata(),c.getReference());
> > +                    analyzer.iterate(q);
> > +                    node.getData().getConjuncts().add(new Conjunct(q,
> > +                                                                   analyzer.getFieldInfo(),
> > +                                                                   c.getReference()));
> >                      break;
> >
> >                  case 2:
> >                      // There are two or more entities referred to in the conjunct
> >                      // This clause can be associated with an edge
>
> // There are two or more entities referred to in the conjunct
>
> Then the comment for case 2 is confusing as the 'or more' is not handled
> by that case.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/608/files/17418539ee2bfbf01e0fd0d96b8e9984280a6bbb#r57705027>
>